// @gear-js/api/cjs/wasm/WasmMeta.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./utils.js');
var importObj = require('./importObj.js');

async function getWasmMetadata(wasmBytes, showDebug = false) {
    const memory = new WebAssembly.Memory({ initial: 256 });
    const module = await WebAssembly.instantiate(wasmBytes, importObj["default"](memory, showDebug));
    const { exports } = module.instance;
    return exports
        ? {
            types: `0x${utils.readMetaValue(memory, exports.meta_registry)}`,
            init_input: utils.readMetaValue(memory, exports.meta_init_input),
            init_output: utils.readMetaValue(memory, exports.meta_init_output),
            async_init_input: utils.readMetaValue(memory, exports.meta_async_init_input),
            async_init_output: utils.readMetaValue(memory, exports.meta_async_init_output),
            handle_input: utils.readMetaValue(memory, exports.meta_handle_input),
            handle_output: utils.readMetaValue(memory, exports.meta_handle_output),
            async_handle_input: utils.readMetaValue(memory, exports.meta_async_handle_input),
            async_handle_output: utils.readMetaValue(memory, exports.meta_async_handle_output),
            title: utils.readMetaValue(memory, exports.meta_title),
            meta_state_input: utils.readMetaValue(memory, exports.meta_state_input),
            meta_state_output: utils.readMetaValue(memory, exports.meta_state_output),
        }
        : {};
}

exports.getWasmMetadata = getWasmMetadata;
