// @gear-js/api/cjs/create-type/generate.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@polkadot/util');
require('@polkadot/util-crypto');
require('@polkadot/types');
var json = require('../utils/json.js');
require('../utils/prefixes.js');
require('../utils/separator.js');
require('@polkadot/api');

const getTree = (type, name, value, count) => {
    const tree = { type, name, value: json.isJSON(value) ? json.toJSON(value) : value };
    return type === 'Array' ? { ...tree, count } : tree;
};
var generate = {
    Primitive: (type) => {
        return getTree('Primitive', type, type);
    },
    Null: () => {
        return getTree('Null', 'Null', 'Null');
    },
    Tuple: (name, type) => {
        return getTree('Tuple', name, type);
    },
    Struct: (name, value) => {
        return getTree('Struct', name, value);
    },
    Enum: (name, options) => {
        return getTree('Enum', name, options);
    },
    Array: (name, type, count) => {
        return getTree('Array', name, type, count);
    },
    Option: (name, some) => {
        return getTree('Option', name, some);
    },
    Result: (name, ok, err) => {
        return getTree('Result', name, { ok, err });
    },
    Vec: (name, type) => {
        return getTree('Vec', name, type);
    },
    BTreeMap: (name, key, value) => {
        return getTree('BTreeMap', name, { key, value });
    },
    BTreeSet: (name, type) => {
        return getTree('BTreeSet', name, type);
    },
};

exports["default"] = generate;
