// @gear-js/api/create-type/generate.js
import '@polkadot/util';
import '@polkadot/util-crypto';
import '@polkadot/types';
import { isJSON, toJSON } from '../utils/json.js';
import '../utils/prefixes.js';
import '../utils/separator.js';
import '@polkadot/api';

const getTree = (type, name, value, count) => {
    const tree = { type, name, value: isJSON(value) ? toJSON(value) : value };
    return type === 'Array' ? { ...tree, count } : tree;
};
var generate = {
    Primitive: (type) => {
        return getTree('Primitive', type, type);
    },
    Null: () => {
        return getTree('Null', 'Null', 'Null');
    },
    Tuple: (name, type) => {
        return getTree('Tuple', name, type);
    },
    Struct: (name, value) => {
        return getTree('Struct', name, value);
    },
    Enum: (name, options) => {
        return getTree('Enum', name, options);
    },
    Array: (name, type, count) => {
        return getTree('Array', name, type, count);
    },
    Option: (name, some) => {
        return getTree('Option', name, some);
    },
    Result: (name, ok, err) => {
        return getTree('Result', name, { ok, err });
    },
    Vec: (name, type) => {
        return getTree('Vec', name, type);
    },
    BTreeMap: (name, key, value) => {
        return getTree('BTreeMap', name, { key, value });
    },
    BTreeSet: (name, type) => {
        return getTree('BTreeSet', name, type);
    },
};

export { generate as default };
